---
title: "Simulation analysis of batching in RNA-seq"
subtitle: ''
author: "Developed by [Gabriel Hoffman](http://gabrielhoffman.github.io/)"
date: "Run on `r Sys.time()`"
documentclass: article
output: 
  html_document:
  toc: true
  smart: false
vignette: >
  %\VignetteIndexEntry{Gene set enrichment from genomic intervals}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
  %\usepackage[utf8]{inputenc}
---


<!---

cd /hpc/users/hoffmg01/work/misc_vp
ml git pandoc
git pull
R


rmarkdown::render("batch_effects_with_overlap.Rmd", output_dir='./', intermediates_dir='./')



 # rm -fr batch_effects_with_overlap_*

# ```{r}
# knitr::knit_exit()
# ```


--->


## TODO

- write code to add time point
- sim fewer reads for shared controls
- what happens if shared is only controls and they are very different
- disease effect is time point dependent
- number of shared controls


Combine study design with statistical modelling to account for batch effects

- Naive: no redundant samples and ignore batch effect
- Simple: no redundant samples, account for batch effect by setting the mean of each batch to be equal for each gene. 
- Replicates: include a set of samples in multiple batches
	- Perform differential expression *jointly* modeling batch, disease, and individual replicates
	- Two step: statistically estimate the expression profile for each individual considering batches and replicates.  This method is ideal because it estimates the corrected expression profile and can be used downstream
- In this simulation experiment, I also include the "oracle" scenerio where the batch correction is performed exactly as if provided by an oracle.


is time point really within batch?are we interested in case control or timepoint comparison?



incomplete data
low read depth
fixed number of experiments


```{r load.packages, echo=FALSE, message=FALSE, results='hide'}
suppressPackageStartupMessages({
library(variancePartition)
library(PRROC)
library(BiocParallel)
library(lme4)
library(gridExtra)
})

# set.seed(1)
register(SnowParam(4))
# register(SerialParam())

knitr::opts_chunk$set(
  echo=FALSE,
  warning=FALSE,
  message=FALSE,
  error = FALSE,
  tidy = FALSE,
  cache = TRUE,
  cache.lazy = FALSE,  dev = c("png", "pdf"), 
  fig.width=7, fig.height=7)

options(markdown.HTML.stylesheet = 'css/custom.css')
```

```{r}
# convert mean and variances in parameters of beta distribution
# https://en.wikipedia.org/wiki/Beta_distribution#Parameter_estimation
# there is a limit on the variance given constraints of the beta distribution.
# If var exceeds this limit, assign maximum allows variance
estimate_beta_parameters = function(mu, v){
	v = min(v, 0.8*mu*(1-mu))

	alpha = mu*(mu*(1-mu)/v - 1)
	beta = (1-mu) *(mu*(1-mu)/v - 1)

	c(alpha=alpha, beta=beta)
}

# Given an expression matrix Y with genes in rows, 
# get mean expression for each individual
mean_per_sampleID = function(Y1, info){

	sid_array = as.character(unique(info$SampleID))
	M = lapply(sid_array, function(SID){
		idx = info[info$SampleID == SID,'idx']
		i = colnames(Y1) %in% as.character(idx)
		m = rowMeans(Y1[,i,drop=FALSE])
		as.matrix(m)
		} )
	M = do.call(cbind, M)
	colnames(M) = sid_array
	M
}
```


```{r define.simulation.functions}

generateSimulations = function( opt, info ){

	# create design matrix for each variable
	design_ID = model.matrix( ~ 0+Individual,info)
	design_Disease = model.matrix( ~ 0+Disease,info)
	design_TimePoint = model.matrix( ~ 0+TimePoint,info)
	design_Batch = model.matrix( ~ 0+Batch,info)

	# Simulate expression data
	simResults = lapply( 1:opt$p, function(j){

		# Individual
		eta_ID = design_ID %*% rnorm(nlevels(info$Individual))

		# Batch
		eta_batch = design_Batch %*% rnorm(nlevels(info$Batch))

		# TimePoint
		eta_TimePoint = design_TimePoint %*% rnorm(nlevels(info$TimePoint))

		# draw variance fractions
		sigSq_ID = rbeta(1, opt$distr_ID[1], opt$distr_ID[2])
		sigSq_Batch = rbeta(1, opt$distr_Batch[1], opt$distr_Batch[2])
		sigSq_TimePoint = rbeta(1, opt$distr_Timepoint[1], opt$distr_Timepoint[2])

		if( ! paste0('gene_', j) %in% opt$tpGenes ){
			sigSq_TimePoint = 0
		}

		# of each sample is observed once, no donor component
		if(  max(table(info$Individual)) == 1){
			sigSq_ID = 1e-4
		}

		# each individual has its own error variance
		v_var = rbeta( nlevels(info$Individual), 1, 1) + 0.5
		resid_var = model.matrix( ~ 0+Individual, info) %*% v_var

		if( paste0('gene_', j) %in% opt$deGenes ){
			# Disease
			eta_Disease = design_Disease %*% rnorm(nlevels(info$Disease))

			sigSq_Disease = rbeta(1, opt$distr_Disease[1], opt$distr_Disease[2])
			sigSq_Resid = max(1 - sigSq_ID - sigSq_Disease - sigSq_Batch - sigSq_TimePoint, .05)

			# combine
			y = scale(eta_ID) * (sigSq_ID-sigSq_Disease) + 
				scale(eta_batch) * sigSq_Batch + 
				scale(eta_Disease) * sigSq_Disease +	
				scale(eta_TimePoint) * sigSq_TimePoint +			 
				rnorm(nrow(info), 0, sd=sqrt(resid_var)) * sigSq_Resid
		}else{
			sigSq_Resid = max(1 - sigSq_ID - sigSq_Batch - sigSq_TimePoint, .05)

			# combine
			y = scale(eta_ID) * sigSq_ID + 
				scale(eta_batch) * sigSq_Batch + 	
				scale(eta_TimePoint) * sigSq_TimePoint +	 
				rnorm(nrow(info), sd=sqrt(resid_var)) * sigSq_Resid
		}
		
		# Perfect correction for batch
		y_oracle = y - scale(eta_batch) * sigSq_Batch

		list(	y 		= t(y),
				y_oracle = t(y_oracle))
	})

	# observed expression data
	Y = do.call(rbind, lapply(simResults, function(x) x$y))
	rownames(Y) = paste0("gene_", 1:nrow(Y))

	# Oracle, with *perfect* batch correction
	Y_oracle = do.call(rbind, lapply(simResults, function(x) x$y_oracle))
	rownames(Y_oracle) = paste0("gene_", 1:nrow(Y_oracle))

	# Get metadata at the individal level.  info is at the experiment level
	metadata = unique(with(info, data.frame(Individual, TimePoint, Disease, stringsAsFactors=FALSE)))
	rownames(metadata) = with(metadata, paste(Individual, TimePoint, sep=':'))
	metadata$Batch = "batch1"

	unique_ID_idx = !duplicated(info$SampleID)

	list(	Y 				= Y,
			Y_oracle 		= Y_oracle,
			metadata		= metadata,
			unique_ID_idx 	= unique_ID_idx,
			deGenes			= opt$deGenes,
			tpGenes			= opt$tpGenes)
}


createInfo = function( opt ){

	n_batches = with(opt, n_total / n_samples_per_batch)
	opt$n_cases = with(opt, (n_total - n_batches*n_controls_per_batch) / n_time_points )

	# Generate study design
	info = data.frame(	Individual = rep(paste0("ID", 1:opt$n_cases), opt$n_time_points),
						TimePoint = unlist(lapply(1:opt$n_time_points, function(x) rep(paste0("time", x), opt$n_cases))),
						Disease = "Case", 
					stringsAsFactors=FALSE)

	# alpha = .5: all equal
	# alpha = 1 very different
	p1 = seq(opt$alpha,1-opt$alpha, length.out=n_batches)
	prob = rbind(p1, p2=1-p1)
	prob = prob / rowSums(prob)
	rownames(prob) = paste0("time", 1:nrow(prob))
	colnames(prob) = paste0("batch", 1:ncol(prob))

	info$Batch = NA
	for( tp in unique(info$TimePoint) ){
		i = which(info$TimePoint == tp)
		info$Batch[i] = paste0("batch", sample.int(n_batches, length(i), replace=TRUE, prob[tp,]))
	}
	info$Batch = factor(info$Batch, paste0('batch', 1:n_batches))
	tab = table(info$Batch)
	targetCount = ceiling(nrow(info) / n_batches)

	# if batches have too many counts, re-allocaite them to other batches
	while( max(tab) > targetCount ){

		# get batch with max counts
		btch = names(tab)[which.max(tab)]

		# sample a set to switch batches
		i = sample(which(info$Batch == btch), max(tab)-targetCount)

		# draw new batch assignments from batches which are too small
		batch_small = names(tab)[which(tab < targetCount)]

		info$Batch[i] = sample(batch_small, length(i), replace=TRUE, prob[tp,batch_small])	
		tab = table(info$Batch)
	}

	# add a set of shared controls to each batch
	infoControl.tmp = data.frame( Individual = paste0("ID", (opt$n_cases+1):(opt$n_cases+opt$n_control)),
		TimePoint = "time1",
		Disease = "control", stringsAsFactors=FALSE)

	# only retain a subset of controls per batch
	infoControl = c()
	for( batch in unique(info$Batch) ){
		i = sort(sample.int( opt$n_control, opt$n_controls_per_batch ))
		infoControl.tmp$Batch = batch
		infoControl = rbind(infoControl, infoControl.tmp[i,])
	}

	# combine cases and controls
	info = rbind(info, infoControl)

	info$SampleID = with(info, paste(Individual, TimePoint, sep=':'))
	info$idx = 1:nrow(info)

	info$Status = with(info, paste(Disease, TimePoint, sep=' - '))

	# set as factors
	info$Individual = factor(info$Individual, unique(info$Individual))
	info$Disease = factor(info$Disease)
	info$Batch = factor(info$Batch, colnames(prob))
	info$TimePoint = factor(info$TimePoint, unique(info$TimePoint))
	info$Status = factor(info$Status)

	info
}

plotStudyDesign = function(info){
	# plot study design
	ratio = with(info, length(unique(Batch))/ length(unique(Individual)) )
	ggplot(info, aes(Individual, fill=Status, y=Batch)) + geom_tile() + theme_bw() + theme(aspect.ratio=ratio, legend.position="bottom", plot.title = element_text(hjust = 0.5), axis.text.x=element_blank()) + ggtitle("Sample batching")
}


evalBatchCorrection = function( simObj, info ){
	
	Y = simObj$Y

	Y_list = list()

	Y_list[["remove_batch_full_model"]] = tryCatch({
		# This gave a strange error in multithreaded mode
		fitList = fitVarPartModel( Y, ~ (1|Batch) + (1|TimePoint) + (1|Disease) + (1|Individual), info, fxn = function(fit){ 
				B = predict(fit, info[,c("Batch"),drop=FALSE], re.form=~(1|Batch))
				fit@resp$y - B
			}, showWarnings=FALSE, BPPARAM=SerialParam(), quiet=TRUE)

		Y_corrected = do.call(rbind, fitList)
		rownames(Y_corrected) = rownames(Y)
		 Y_corrected
	}, error = function(e) NULL )

	Y_list[["remove_batch_simple"]] = removeBatchEffect(Y, info$Batch, design=model.matrix(~TimePoint, info))

	Y_list
}


evalDifferentialExpression = function(simObj, info, Y_list, test = "TimePoint", corrRetain=.9){

	Y = simObj$Y
	Y_oracle = simObj$Y_oracle

	suppressWarnings({
	fitList = list()
	idx = match(colnames(Y_oracle), info$idx)
	form = switch(test, "TimePoint" = ~ (1|Disease) + TimePoint + (1|Individual),
						"Disease" 	= ~ Disease + (1|TimePoint) + (1|Individual))
	# form = switch(test, "TimePoint" = ~ Disease + TimePoint,# + (1|Individual),
	# 					"Disease" 	= ~ Disease + TimePoint)# + (1|Individual))
	fitList[['Oracle']] = tryCatch({dream(Y_oracle, form, info[idx,], quiet=TRUE)}, error = function(e) NULL )

	if( is.null(fitList[['Oracle']]) ){
		idx = match(colnames(Y_oracle), info$idx)
		form = switch(test, "TimePoint" = ~ TimePoint + (1|Individual),
							"Disease" 	= ~ Disease + (1|TimePoint) + (1|Individual))
		# form = switch(test, "TimePoint" = ~ Disease + TimePoint,# + (1|Individual),
		# 					"Disease" 	= ~ Disease + TimePoint)# + (1|Individual))
		fitList[['Oracle']] = tryCatch({dream(Y_oracle, form, info[idx,], quiet=TRUE)}, error = function(e) NULL )

	}

	form = switch(test, "TimePoint" = ~ (1|Disease) + (1|Batch) + TimePoint + (1|Individual),
						"Disease" 	= ~ Disease + (1|Batch) + (1|TimePoint) + (1|Individual))
	fitList[['Replicates - joint model']] = tryCatch({dream(Y, form, info, quiet=TRUE)}, error = function(e) NULL )

	for(v in corrRetain){
		# add noise just to controls
		# construct so that cor(Y[j,], Y_noise[j,])^2 = corrRetain
		Y_noise = Y
		idx = which(info$Disease == "control")
		for(j in 1:nrow(Y) ){
			y = Y_noise[j,idx]
			Y_noise[j,idx] = y + rnorm(length(y), 0, sd=sqrt(var(y)*(1-v)/v)) 
		}
		# j=10
		# cor(Y[j,idx], Y_noise[j,idx])^2
		# corrRetain

		key = paste0('Replicates - joint model (R^2=',v, ')')

		form = switch(test, "TimePoint" = ~ (1|Disease) + (1|Batch) + TimePoint + (1|Individual),
							"Disease" 	= ~ Disease + (1|Batch) + (1|TimePoint) + (1|Individual))
		fitList[[key]] = tryCatch({dream(Y_noise, form, info, quiet=TRUE)}, error = function(e) NULL )

	}

	form = switch(test, "TimePoint" = ~ (1|Disease) + TimePoint + (1|Individual),
						"Disease" 	= ~ Disease + (1|TimePoint) + (1|Individual))
	fitList[['Replicates: Two step']] = tryCatch({dream(Y_list[["remove_batch_full_model"]], form, info, quiet=TRUE)}, error = function(e) NULL )

	if( test == "TimePoint"){
		idx = which(info$Disease == "Case")
		form = ~ (1|Batch) + TimePoint + (1|Individual)
		fitList[['Simple: No controls, no replicates (random)']] = tryCatch({dream(Y[,idx], form, info[idx,], quiet=TRUE)}, error = function(e) NULL )

		idx = which(info$Disease == "Case")
		form = ~ Batch + TimePoint + Individual
		fitList[['Simple: No controls, no replicates (fixed)']] = tryCatch({eBayes(capture.output(dream(Y[,idx], form, info[idx,], quiet=TRUE)))}, error = function(e) NULL )
		
		idx = which(info$Disease == "Case")
		form = ~ TimePoint + Individual
		fitList[['Naive: No controls, no replicates, ignore batch (fixed)']] = tryCatch({eBayes(capture.output(dream(Y[,idx], form, info[idx,], quiet=TRUE)))}, error = function(e) NULL )

	}else{
		form = ~ (1|Batch) + (1|TimePoint) + Disease + (1|Individual)
		fitList[['Simple']] = tryCatch({dream(Y, form, info, quiet=TRUE)}, error = function(e) NULL )

		
		# form = ~ Batch + (1|TimePoint) + Disease + (1|Individual)
		# fitList[['Simple: - fixed']] = tryCatch({dream(Y, form, info, quiet=TRUE)}, error = function(e) NULL )
	}
	})

	fitList
}


evalPerformance = function( fitList, coef, positives){

	res = lapply(names(fitList), function(method){

		tab = topTable(fitList[[method]], coef=coef, number=Inf, sort.by="none")

		# add missing genes to tab with P.Value = 1
		# excluded = which(!rownames(Y) %in% rownames(tab))

		# if( length(excluded) > 0){
		# 	tab_excluded = tab[1:length(excluded),]
		# 	rownames(tab_excluded) = rownames(Y)[excluded]
		# 	tab_excluded[] = NA
		# 	tab_excluded$P.Value = 1
		# 	tab = rbind(tab, tab_excluded)
		# }

		# AUPR
		idx = rownames(tab) %in% positives
		pr <- pr.curve( -log10(tab$P.Value[idx]), -log10(tab$P.Value[!idx]), curve=TRUE, rand.compute=TRUE )

		# False positive rate
		FPR = sum(tab$P.Value[!idx] < 0.05) / length(tab$P.Value[!idx])

		data.frame( method 		= method, 
					AUPR 		= pr$auc.integral, 
					AUPR.random = pr$rand$auc.integral,
					FPR 		= FPR,
					stringsAsFactors=FALSE)
		})
	res = do.call(rbind, res)

	# ord = c('Naive: No controls, no replicates, ignore batch',
	# 		'Simple: No controls, no replicates',
	# 		'Replicates: Two step',
	# 		'Replicates - joint model + noiseToControls',
	# 		'Replicates - joint model',
	# 		'Oracle') 
 
	# res$method = factor(res$method, ord)

	res
}

shifter <- function(x, n = 1) {
  if (n == 0) x else c(tail(x, -n), head(x, n))
}

plotStratify = function (formula, data, xlab, ylab, main, sortBy, colorBy, sort = TRUE,
    text = NULL, text.y = 1, text.size = 5, pts.cex = 1, ylim = NULL,
    legend = TRUE, x.labels = FALSE){
    mc <- match.call()
    m <- match(c("formula", "data"), names(mc), 0L)
    mf <- mc[c(1L, m)]
    mf[[1L]] <- as.name("model.frame")
    mf <- eval(mf, parent.frame())
    data.st <- data.frame(mf)
    if (ncol(data.st) != 2) {
        stop("formula must have exactly 2 entries")
    }
    xval = colnames(data.st)[attr(attr(mf, "terms"), "response") +
        1]
    yval = colnames(data.st)[attr(attr(mf, "terms"), "response")]
    if (missing(xlab)) {
        xlab = xval
    }
    if (missing(sortBy) || is.null(sortBy)) {
        sortBy = xval
    }
    if (missing(colorBy)) {
        colorBy = xval
    }
    if (missing(ylab)) {
        ylab = yval
    }
    if (!(sortBy %in% colnames(data.st))) {
        stop(paste("sortBy is not found in colnames(data): sortBy =",
            sortBy))
    }
    data.st[[yval]] = as.numeric(data.st[[yval]])
    xpos = 0.5
    ypos = text.y * (max(data.st[[yval]]) - min(data.st[[yval]])) +
        min(data.st[[yval]])
    if (sort) {
        data.st[["reorder"]] = reorder(data.st[[sortBy]], data.st[[yval]],
            FUN = median)
        ord = "reorder"
    }
    else {
        ord = xval
    }
    pOut = ggplot(data.st, aes_string(x = ord, y = yval)) + theme_bw() +
        theme(plot.background = element_blank(), panel.grid.major = element_blank(),
            panel.grid.minor = element_blank(), panel.background = element_blank()) +
        ylab(ylab) + xlab(xlab) + theme(plot.title = element_text(hjust = 0.5))
    if (any(is.null(colorBy)) || any(is.na(colorBy))) {
        pOut = pOut + geom_boxplot(color = "grey", fill = "grey",
            outlier.colour = "black", outlier.shape = 20)
    }
    else {
        if (sum(levels(data.st[[xval]]) %in% names(colorBy)) ==
            nlevels(data.st[[xval]])) {
            i = match(levels(data.st[[ord]]), levels(data.st[[xval]]))
            pOut = pOut + geom_boxplot(aes_string(fill = xval, middle=mean(yval)),
                color = colorBy[i], outlier.colour = "black",
                outlier.shape = 20) + scale_fill_manual(values = array(colorBy))
        }
        else {
            pOut = pOut + geom_boxplot(aes_string(color = colorBy,
                fill = colorBy), outlier.colour = "black", outlier.shape = 20)
        }
        if (legend) {
            pOut = pOut + theme(legend.justification = c(1, 0),
                legend.position = c(1, 0), legend.key = element_rect(fill = "transparent"),
                axis.text.x = element_text(angle = 30), legend.background = element_rect(fill = "transparent"))
        }
        else {
            pOut = pOut + theme(legend.position = "none", axis.text.x = element_text(angle = 30))
        }
    }
    if (!x.labels) {
        pOut = pOut + theme(axis.ticks.x = element_blank(), axis.text.x = element_blank())
    }
    pOut = pOut + stat_summary(geom = "crossbar", width = 0.65,
        fatten = 0, color = "black", fun.data = function(x) {
            return(c(y = median(x), ymin = median(x), ymax = median(x)))
        })
    if (!missing(ylim)) {
        pOut = pOut + ylim(ylim)
    }
    if (!missing(main)) {
        pOut = pOut + ggtitle(main)
    }
    if (!missing(text)) {
        pOut = pOut + annotate("text", label = text, x = xpos,
            y = ypos, size = text.size, hjust = 0)
    }
    return(pOut)
}
```


```{r simulation.params, fig.width=12, fig.height=4}

opt = list()

opt$n_total = 300 # total number of experiments
opt$n_samples_per_batch = 10 # size of each batch
opt$n_controls_per_batch = 2 # modify this parameter

opt$n_time_points = 2
# total number of experiments: N*n_time_points
opt$p = 400
opt$n_de = 100 # genes different between case and control
opt$n_tp = 100 # genes different between time points
opt$corrRetain = .8 

opt$distr_ID = estimate_beta_parameters( .3, .01)
opt$distr_Disease = estimate_beta_parameters( .3, .01)
opt$distr_Timepoint = estimate_beta_parameters( .08, .02)
opt$distr_Batch = estimate_beta_parameters( .4, .01)

# distribution of batch probabilites
opt$alpha = .1

# get names of differentially expressed genes
opt$deGenes = paste0("gene_", sample.int(opt$p, opt$n_de))
opt$tpGenes = paste0("gene_", sample.int(opt$p, opt$n_tp))
```


```{r run.simulation}
info = createInfo( opt )

simObj = generateSimulations( opt, info )
 
Y_list = evalBatchCorrection( simObj, info )
```

```{r plotStudy, fig.width=15, fig.height=3}
plotStudyDesign( info ) 
```

```{r canCorPairs, fig.width=5, fig.height=5}
suppressWarnings({
C = canCorPairs(~ Individual + TimePoint + Disease + Batch, info)
}) 
plotCorrMatrix( C )
```

# VariancePartition results
```{r variancePartition, fig.width=12, eval=FALSE}
# fit variancePartition model
vp = fitExtractVarPartModel( simObj$Y, ~ (1|TimePoint) + (1|Disease) + (1|Batch) + (1|Individual), info)
fig1 = plotVarPart( vp ) + ggtitle("Observed data")

# fit variancePartition model on clean data with no batch effect
vp_oracle = fitExtractVarPartModel( simObj$Y_oracle, ~ (1|TimePoint) + (1|Disease) + (1|Batch) + (1|Individual), info)
fig2 = plotVarPart( vp_oracle ) + ggtitle("Oracle")

grid.arrange(fig1, fig2, ncol=2)
```


# PCA using multiple types of corrections
Green points are controls.  Red is case - time point 1, black is case - time point 2.
```{r pca, fig.width=8, fig.height=8}
# Plot coloring by individual
par(mfrow=c(2,2), pty='s')
# PCA of data with no batch effect
dcmp = prcomp( t(simObj$Y_oracle), scale=TRUE)
pct = format(dcmp$sdev[1:2]^2 / sum(dcmp$sdev^2)*100, digits=3)
plot(dcmp$x[,1:2], col=info$Status, pch=as.integer(info$Disease)+6, 
	xlab = paste0('PC1 (', pct[1], '%)'), ylab = paste0('PC2 (', pct[1], '%)'),
	main="Oracle")
   
# PCA of original data
dcmp = prcomp( t(simObj$Y), scale=TRUE)
pct = format(dcmp$sdev[1:2]^2 / sum(dcmp$sdev^2)*100, digits=3)
plot(dcmp$x[,1:2], col=info$Status, pch=as.integer(info$Disease)+6, 
	xlab = paste0('PC1 (', pct[1], '%)'), ylab = paste0('PC2 (', pct[1], '%)'),
	main="Observed data")

dcmp = prcomp( t(Y_list[["remove_batch_full_model"]]), scale=TRUE)
pct = format(dcmp$sdev[1:2]^2 / sum(dcmp$sdev^2)*100, digits=3)
plot(dcmp$x[,1:2], col=info$Status, pch=as.integer(info$Disease)+6, 
	xlab = paste0('PC1 (', pct[1], '%)'), ylab = paste0('PC2 (', pct[1], '%)'),
	 main="Full model of replicates")

dcmp = prcomp( t(Y_list[["remove_batch_simple"]]), scale=TRUE)
pct = format(dcmp$sdev[1:2]^2 / sum(dcmp$sdev^2)*100, digits=3)
plot(dcmp$x[,1:2], col=info$Status, pch=as.integer(info$Disease)+6, 
	xlab = paste0('PC1 (', pct[1], '%)'), ylab = paste0('PC2 (', pct[1], '%)'),
	main="Simple batch correction")

```

# Gene expression after correction
```{r expr.residuals}
library(gginnards)

show_mean = function(fig){
	delete_layers(fig, idx=2) + stat_summary(geom = "crossbar", width = 0.65,
        fatten = 0, color = "black", fun.data = function(x) {
            return(c(y = mean(x), ymin = mean(x), ymax = mean(x)))})
}

idx = which(! simObj$Y_oracle %in% c(opt$deGenes, opt$tpGenes))
i = idx[1]

df = data.frame(y = t(t(simObj$Y_oracle[i,])), info, stringsAsFactors=FALSE)
y_oracle = df$y
fig1 = plotStratify( y ~ Batch, df, main="Oracle", sort=FALSE, legend=FALSE)

df = data.frame(y = t(t(simObj$Y[i,])), info)
dfcor = list("Observed data" = cor(y_oracle,df$y))
fig2 = plotStratify( y ~ Batch, df, main="Observed data", sort=FALSE, legend=FALSE)

exprResid = Y_list[['remove_batch_full_model']]
df = data.frame(y = t(t(exprResid[i,])), info)
dfcor[["Full model"]] = cor(y_oracle,df$y)
fig3 = plotStratify( y ~ Batch, df, main="Full model", sort=FALSE, legend=FALSE)

exprResid = Y_list[['remove_batch_simple']]
df = data.frame(y = t(t(exprResid[i,])), info)
dfcor[["Simple model"]] = cor(y_oracle,df$y)
fig4 = plotStratify( y ~ Batch, df, main="Simple model", sort=FALSE, legend=FALSE)

grid.arrange(show_mean(fig1), 
			show_mean(fig2), 
			show_mean(fig3), 
			show_mean(fig4))

dfcor2 = data.frame(test = names(dfcor), Rsq=unlist(dfcor)^2)
ggplot( dfcor2, aes(test, Rsq, fill=test)) + geom_bar(stat="identity") + theme_bw() + theme(aspect.ratio=1) + coord_flip()

# get correlations genome-wdie
dfcor = lapply(1:nrow(simObj$Y_oracle), function(j){

	y_oracle = simObj$Y_oracle[j,]

	data.frame("Observed data" 	= cor(y_oracle,  simObj$Y[j,])^2,
				"Full model" 	= cor(y_oracle, Y_list[['remove_batch_full_model']][j,])^2,
				"Simple model" 	= cor(y_oracle, Y_list[['remove_batch_simple']][j,])^2)

	})
dfcor = do.call(rbind, dfcor)

plotVarPart(dfcor)

```


# Performance for single simulation
```{r differential.expression}

fitList = evalDifferentialExpression(simObj, info, Y_list)
  
res = evalPerformance( fitList[1:2], coef='TimePointtime2', positives = simObj$tpGenes)

ggplot(res, aes(method, AUPR, fill=method)) + geom_bar(stat="identity") + theme_bw() + theme(aspect.ratio=1, plot.title = element_text(hjust = 0.5), legend.position="none") + xlab("Method") + scale_y_continuous(limits=c(0, 1.02), expand=c(0, 0)) + geom_hline(yintercept = res$AUPR.random[1], linetype="dashed") + coord_flip() 

ggplot(res, aes(method, FPR, fill=method)) + geom_bar(stat="identity") + theme_bw() + theme(aspect.ratio=1, plot.title = element_text(hjust = 0.5), legend.position="none") + xlab("Method") + scale_y_continuous(limits=c(0, .5), expand=c(0, 0)) + geom_hline(yintercept = 0.05, linetype="dashed") + coord_flip() 

```

## Compare coexpression
```{r coexpression}
library(Matrix)

coexprNorm = function(Yhat){

	if( is.null(Yhat) ){
		res = NA
	}else{
		C = cor(t(Yhat))
		res = norm(C, "F")
	}
	# sum(eigen(C)$values)

	# n = ncol(Yhat)
	# A = scale(t(Yhat)) / sqrt(n-1)
	# # meff(eigen=eigen(crossprod(A))$values, method="gao")

	# evs = svd(A)$d^2
	# sqrt(sum(evs))
	res
}

```





 # rm -fr batch_effects_with_overlap_*

# ```{r}
# knitr::knit_exit()
# ```


# Iterate through simulations
```{r big.sim, message=FALSE}

opt = list()

opt$n_total = 300#3072 # total number of experiments
# opt$n_samples_per_batch = 10 # size of each batch
# opt$n_controls_per_batch = 2 # modify this parameter
opt$n_time_points = 2
# total number of experiments: N*n_time_points
opt$p = 500
opt$n_de = 250 # genes different between case and control
opt$n_tp = 250 # genes different between time points
opt$corrRetain = .8

opt$distr_ID = estimate_beta_parameters( .3, .01)
opt$distr_Disease = estimate_beta_parameters( .08, .01)
opt$distr_Timepoint = estimate_beta_parameters( .05, .02)
opt$distr_Batch = estimate_beta_parameters( .4, .01)

# distribution of batch probabilites
# opt$alpha = .5

# get names of differentially expressed genes
opt$deGenes = paste0("gene_", sample.int(opt$p, opt$n_de))
opt$tpGenes = paste0("gene_", sample.int(opt$p, opt$n_tp))

# opt$n_controls_per_batch = 2
# plotStudyDesign(createInfo( opt ))



# loop through simulations
resultsAll = lapply( c(seq(0.01, 0.1, length.out=1)), function(alpha){

	message("Generate large dataset")

	opt$alpha = alpha

	# In order to ensure the same simulation parameters are used, 
	# subset this larger dataset
	set.seed(1)   
	# opt$n_total = 1200
	# Use this here, to initialise 
	opt$n_controls_per_batch = 30
	opt$n_samples_per_batch = 30*2
	infoTotal = createInfo( opt )
	simObjTotal = generateSimulations( opt, infoTotal )

	# res = lapply( c(0:5), function(n_controls_per_batch){
	res = lapply( c(3), function(n_controls_per_batch){

		opt$n_controls_per_batch = n_controls_per_batch

		res = lapply( 6, function(n_samples_per_batch){
			opt$n_samples_per_batch = n_samples_per_batch

			res = NULL
			v = with(opt, n_total / n_samples_per_batch)
			if(  v == round(v) ){

				# Subset larger dataset
				#-----
				# controls
				if( opt$n_controls_per_batch > 0){
					i = which(with(infoTotal, Disease == "control"))
					indivs_control = as.character(unique(infoTotal$Individual[i])[1:opt$n_controls_per_batch])
				}else{
					indivs_control = c()
				}

				# cases
				n_cases_per_batch = with(opt, n_samples_per_batch - n_controls_per_batch)	
				i = which(with(infoTotal, Disease == "Case"))
				indivs_case = as.character(unique(infoTotal$Individual[i])[1:n_cases_per_batch])

				# subset
				idx = with(infoTotal, which(Individual %in% c(indivs_control, indivs_case)))

				info = infoTotal[idx,]
				simObj = with(simObjTotal, list(Y = Y[,idx], Y_oracle = Y_oracle[,idx]))

				Y_list = evalBatchCorrection( simObj, info )

				message(opt$n_controls_per_batch, ' ', opt$n_samples_per_batch )

				fitListTP = evalDifferentialExpression(simObj, info, Y_list, test="TimePoint", corrRetain=c(0.8, .9))
				resTP = evalPerformance( fitListTP, coef='TimePointtime2', positives = opt$tpGenes)
				resTP$test = "TimePoint"

				message("\tDE Disease")
				fitListDisease = evalDifferentialExpression(simObj, info, Y_list, test="Disease", corrRetain=c( .8, .9))
				resDisease = evalPerformance( fitListDisease, coef='Diseasecontrol', positives = opt$deGenes)	
				resDisease$test = "Disease"

				if( length(resDisease$test) ==1){
					res = resTP
				}else{
					res = rbind(resTP, resDisease)
				}
				res$n_control = opt$n_control
				res$n_samples_per_batch = opt$n_samples_per_batch
				res$n_controls_per_batch = opt$n_controls_per_batch
				res$alpha = opt$alpha

				# Evaluate coexpression norm
				# cpn = c(coexprNorm( Y_list ), coexprNorm( simObj) )
				# res$coexprNorm = paste(sapply( names(cpn), function(x) paste0(x, ':', cpn[x], ';')), collapse='')
				# browser()
				# save(list=ls(), file="test.RDATA")

				cpn = c( 	Oracle 				= coexprNorm( simObj[['Y_oracle']] ),
							'Replicates - joint model' = coexprNorm( Y_list[['remove_batch_full_model']] ),
							'Simple: No controls, no replicates (random)' = coexprNorm( Y_list[['remove_batch_simple']] ) )
			
				h = match(res$method[res$test=="TimePoint"], names(cpn))
				res$coexprNorm = c(NA)
				res$coexprNorm[res$test=="TimePoint"][h[!is.na(h)]] = cpn
			}
			res
		})
		do.call('rbind', res)     
	})
	do.call('rbind', res)    
})
resultsAll = do.call('rbind', resultsAll)

```

```{r study.design}

opt$alpha = .3
# loop through simulations
resStudy = lapply( c(1:8), function(n_controls_per_batch){
	opt$n_controls_per_batch = n_controls_per_batch

	res = lapply( seq(20, 100, by=2), function(n_samples_per_batch){
		opt$n_samples_per_batch = n_samples_per_batch

		df = NULL
		v =  with(opt, n_total / n_samples_per_batch)
		if(  v == round(v) ){

			# cat( opt$n_controls_per_batch, ' ', opt$n_samples_per_batch, "\n")

			info = createInfo( opt )

			df = data.frame( 	n_controls_per_batch= opt$n_controls_per_batch, 
								n_samples_per_batch = opt$n_samples_per_batch, 
								n_cases 			= sum(info$Disease=="Case"))
		}
		df
	})
	do.call(rbind, res)
})
resStudy = do.call(rbind, resStudy)

ggplot(resStudy, aes(n_controls_per_batch, n_cases, color=as.character(n_samples_per_batch))) + geom_line() + theme_bw() + theme(aspect.ratio=1, plot.title = element_text(hjust = 0.5)) + ylab("Number of cases") + xlab("Number of controls per batch") + scale_color_discrete("Number of samples\nper batch") + ggtitle("Balancing cases and controls in study design")
```






```{r plot.results, fig.width=10, cache=FALSE, fig.height=12}

#   
df = resultsAll
# df = df[grep("R\\^2", df$method, invert=TRUE),]

aupr.random.TimePoint = unique(df[df$test=="TimePoint", 'AUPR.random'])
aupr.random.Disease = unique(df[df$test=="Disease", 'AUPR.random'])

# pdf("~/www/test.pdf", width=20)
ggplot(subset(df, test=="TimePoint"), aes(n_controls_per_batch, AUPR, color=method)) + geom_line(size=2) + facet_wrap(~alpha + n_samples_per_batch, nrow=1) + theme_bw() + theme(aspect.ratio=1, plot.title = element_text(hjust = 0.5)) + ggtitle("Test TimePoint") 	#+ geom_hline(yintercept=aupr.random.TimePoint, linetype='dashed')
# dev.off()




ggplot(subset(df, test=="TimePoint"), aes(n_control, FPR, color=method)) + geom_line(size=2) + facet_wrap(~alpha + n_samples_per_batch, nrow=1) + theme_bw() + theme(aspect.ratio=1, plot.title = element_text(hjust = 0.5)) + ggtitle("Test TimePoint") + geom_hline(yintercept=0.05, linetype='dashed') + ylim(0, NA)


ggplot(subset(df, test=="Disease"), aes(n_control, AUPR, color=method)) + geom_line(size=2) + facet_wrap(~alpha, nrow=1) + theme_bw() + theme(aspect.ratio=1, plot.title = element_text(hjust = 0.5)) + ggtitle("Test Disease") + geom_hline(yintercept=aupr.random.Disease, linetype='dashed')


```


## Coexpression
In this simulation all genes are statistically independent, so the true correlation matrix is `diag(1,p)`.  Here, I evaluate the estimated correlation matrix and the distance to the true correlation matrix using the [Frobenious norm](https://en.wikipedia.org/wiki/Matrix_norm#Frobenius_norm).

```{r eval.coexprNorm}

df_coexprNorm = 




```

How many shared controls is useful? vs noise add disease?


how many overlaps between batches
	do we need a shared set, or some overlaps










