---
title: "Simulation analysis of batching in RNA-seq"
subtitle: ''
author: "Developed by [Gabriel Hoffman](http://gabrielhoffman.github.io/)"
date: "Run on `r Sys.time()`"
documentclass: article
output: 
  html_document:
  toc: true
  smart: false
vignette: >
  %\VignetteIndexEntry{Gene set enrichment from genomic intervals}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
  %\usepackage[utf8]{inputenc}
---


<!---


 rmarkdown::render("batch_effects_with_overlap.Rmd", output_dir='./', intermediates_dir='./')


# ```{r}
# knitr::knit_exit()
# ```


--->


## TODO

- write code to add time point
- sim fewer reads for shared controls
- what happens if shared is only controls and they are very different
- disease effect is time point dependent


Combine study design with statistical modelling to account for batch effects

- Naive: no redundant samples and ignore batch effect
- Simple: no redundant samples, account for batch effect by setting the mean of each batch to be equal for each gene. 
- Replicates: include a set of samples in multiple batches
	- Perform differential expression *jointly* modeling batch, disease, and individual replicates
	- Two step: statistically estimate the expression profile for each individual considering batches and replicates.  This method is ideal because it estimates the corrected expression profile and can be used downstream
- In this simulation experiment, I also include the "oracle" scenerio where the batch correction is performed exactly as if provided by an oracle.


is time point really within batch?
are we interested in case control or timepoint comparison?


```{r load.packages, echo=FALSE, message=FALSE, results='hide'}
suppressPackageStartupMessages({
library(variancePartition)
library(PRROC)
library(BiocParallel)
register(SerialParam())
library(lme4)
library(BiocParallel)
library(gridExtra)
})

set.seed(1)
register(SnowParam(4))

knitr::opts_chunk$set(
  echo=FALSE,
  warning=FALSE,
  message=FALSE,
  error = FALSE,
  tidy = FALSE,
  cache = FALSE,
  cache.lazy = FALSE,  dev = c("png", "pdf"), 
  fig.width=7, fig.height=7)

options(markdown.HTML.stylesheet = 'css/custom.css')
```

```{r}
# convert mean and variances in parameters of beta distribution
# https://en.wikipedia.org/wiki/Beta_distribution#Parameter_estimation
# there is a limit on the variance given constraints of the beta distribution.
# If var exceeds this limit, assign maximum allows variance
estimate_beta_parameters = function(mu, v){
	v = min(v, 0.8*mu*(1-mu))

	alpha = mu*(mu*(1-mu)/v - 1)
	beta = (1-mu) *(mu*(1-mu)/v - 1)

	c(alpha=alpha, beta=beta)
}

# Given an expression matrix Y with genes in rows, 
# get mean expression for each individual
mean_per_sampleID = function(Y1, info){

	sid_array = as.character(unique(info$SampleID))
	M = lapply(sid_array, function(SID){
		idx = info[info$SampleID == SID,'idx']
		i = colnames(Y1) %in% as.character(idx)
		m = rowMeans(Y1[,i,drop=FALSE])
		as.matrix(m)
		} )
	M = do.call(cbind, M)
	colnames(M) = sid_array
	M
}
```

```{r simulation.params, fig.width=12, fig.height=4}

n_cases = 100 # total number of individuals
n_control = 10 # number of control individuals
n_time_points = 2
# total number of experiments: N*n_time_points
p = 500
n_de = 200 # genes different between case and control
n_tp = 250 # genes different between time points
n_samples_per_batch = 20
n_batches = n_cases*n_time_points / n_samples_per_batch
n_batches

# n_batches_per_timePoint = n_cases / n_samples_per_batch


opt = list()
opt$distr_ID = estimate_beta_parameters( .3, .05)
opt$distr_Disease = estimate_beta_parameters( .2, .05)
opt$distr_Timepoint = estimate_beta_parameters( .3, .1)
opt$distr_Batch = estimate_beta_parameters( .3, .05)

# Generate study design
info = data.frame(	Individual = rep(paste0("ID", 1:n_cases), n_time_points),
					TimePoint = unlist(lapply(1:n_time_points, function(x) rep(paste0("time", x), n_cases))),
					Disease = "Case", 
				stringsAsFactors=FALSE)

# since samples from the sample time point are all in the same batch, 
# include the timepoint name in the batch name
# info$Batch = NA
# for( tp in unique(info$TimePoint) ){
# 	i = which(info$TimePoint == tp)
# 	info$Batch[i] = paste0(tp, "_Batch", sample.int(n_batches_per_timePoint, length(i), replace=TRUE))
# }

# For a given probability of being in each batch
# prob = rbind(	c(.5, .2, .2, .1, .3),
				# c(.2, .5, .2, .1, .3))
prob = rbind(	rep(1/n_batches, n_batches),
				rep(1/n_batches, n_batches))
prob = prob / rowSums(prob)
rownames(prob) = paste0("time", 1:nrow(prob))
colnames(prob) = paste0("batch", 1:ncol(prob))

info$Batch = NA
for( tp in unique(info$TimePoint) ){
	i = which(info$TimePoint == tp)
	info$Batch[i] = paste0("batch", sample.int(n_batches, length(i), replace=TRUE, prob[tp,]))
}
table(info$Batch)



# add a set of shared controls to each batch
infoControl.tmp = data.frame( Individual = paste0("ID", (n_cases+1):(n_cases+n_control)),
	TimePoint = "time0",
	Disease = "control", stringsAsFactors=FALSE)

infoControl = c()
for( batch in unique(info$Batch) ){
	infoControl.tmp$Batch = batch
	infoControl = rbind(infoControl, infoControl.tmp)
}

# combine cases and controls
info = rbind(info, infoControl)

info$SampleID = with(info, paste(Individual, TimePoint, sep=':'))
info$idx = 1:nrow(info)

# set as factors
info$Individual = factor(info$Individual, unique(info$Individual))
info$Disease = factor(info$Disease)
info$Batch = factor(info$Batch, colnames(prob))
info$TimePoint = factor(info$TimePoint, unique(info$TimePoint))


ratio = with(info, length(unique(Batch))/ length(unique(Individual)) )
ggplot(info, aes(Individual, fill=TimePoint, y=Batch)) + geom_tile() + theme_bw() + theme(aspect.ratio=ratio, legend.position="bottom", plot.title = element_text(hjust = 0.5), axis.text.x=element_blank()) + ggtitle("Sample batching")


# create design matrix for each variable
design_ID = model.matrix( ~ 0+Individual,info)
design_Disease = model.matrix( ~ 0+Disease,info)
design_TimePoint = model.matrix( ~ 0+TimePoint,info)
design_Batch = model.matrix( ~ 0+Batch,info)

# get names of differentially expressed genes
deGenes = paste0("gene_", 1:n_de)
tpGenes = paste0("gene_", sample.int( p, n_tp))

# Simulate expression data
simResults = lapply( 1:p, function(j){

	# Individual
	eta_ID = design_ID %*% rnorm(nlevels(info$Individual))

	# Batch
	eta_batch = design_Batch %*% rnorm(nlevels(info$Batch))

	# TimePoint
	eta_TimePoint = design_TimePoint %*% rnorm(nlevels(info$TimePoint))

	# draw variance fractions
	sigSq_ID = rbeta(1, opt$distr_ID[1], opt$distr_ID[2])
	sigSq_Batch = rbeta(1, opt$distr_Batch[1], opt$distr_Batch[2])
	sigSq_TimePoint = rbeta(1, opt$distr_Timepoint[1], opt$distr_Timepoint[2])

	if( ! paste0('gene_', j) %in% tpGenes ){
		sigSq_TimePoint = 0
	}

	# of each sample is observed once, no donor component
	if(  max(table(info$Individual)) == 1){
		sigSq_ID = 1e-4
	}

	# each individual has its own error variance
	v_var = rbeta( nlevels(info$Individual), 1, 1) + 0.5
	resid_var = model.matrix( ~ 0+Individual, info) %*% v_var

	if( j <= n_de){
		# Disease
		eta_Disease = design_Disease %*% rnorm(nlevels(info$Disease))

		sigSq_Disease = rbeta(1, opt$distr_Disease[1], opt$distr_Disease[2])
		sigSq_Resid = max(1 - sigSq_ID - sigSq_Disease - sigSq_Batch - sigSq_TimePoint, .05)

		# combine
		y = scale(eta_ID) * (sigSq_ID-sigSq_Disease) + 
			scale(eta_batch) * sigSq_Batch + 
			scale(eta_Disease) * sigSq_Disease +	
			scale(eta_TimePoint) * sigSq_TimePoint +			 
			rnorm(nrow(info), 0, sd=sqrt(resid_var)) * sigSq_Resid
	}else{
		sigSq_Resid = max(1 - sigSq_ID - sigSq_Batch - sigSq_TimePoint, .05)

		# combine
		y = scale(eta_ID) * sigSq_ID + 
			scale(eta_batch) * sigSq_Batch + 	
			scale(eta_TimePoint) * sigSq_TimePoint +	 
			rnorm(nrow(info), sd=sqrt(resid_var)) * sigSq_Resid
	}
	
	y_clean = y - scale(eta_batch) * sigSq_Batch

	list(y = t(y),
		y_clean = t(y_clean))
})

# observed expression data
Y = do.call(rbind, lapply(simResults, function(x) x$y))
rownames(Y) = paste0("gene_", 1:nrow(Y))

# with no batch effect
Y_clean = do.call(rbind, lapply(simResults, function(x) x$y_clean))
rownames(Y_clean) = paste0("gene_", 1:nrow(Y_clean))

# Get metadata at the individal level.  info is at the experiment level
metadata = unique(with(info, data.frame(Individual, TimePoint, Disease, stringsAsFactors=FALSE)))
rownames(metadata) = with(metadata, paste(Individual, TimePoint, sep=':'))

unique_ID_idx = !duplicated(info$SampleID)

```

# VariancePartition results
```{r variancePartition, fig.width=12}
# fit variancePartition model
vp = fitExtractVarPartModel( Y, ~ (1|TimePoint) + (1|Disease) + (1|Batch) + (1|Individual), info)
fig1 = plotVarPart( vp ) + ggtitle("Observed data")

# fit variancePartition model on clean data with no batch effect
vp_clean = fitExtractVarPartModel( Y_clean, ~ (1|TimePoint) + (1|Disease) + (1|Batch) + (1|Individual), info)
fig2 = plotVarPart( vp_clean ) + ggtitle("Oracle")

grid.arrange(fig1, fig2, ncol=2)
```

```{r batch.correction}

Y_list = list()

# Simpler variancePartition call
fitList = fitVarPartModel( Y, ~ (1|Batch) + (1|Individual:TimePoint), info, fxn = function(fit){ 
		t(lme4::ranef(fit)$'Individual:TimePoint') + as.numeric(lme4::fixef(fit)[1]) 
	}, showWarnings=FALSE)

Y_corrected = do.call(rbind, fitList)
rownames(Y_corrected) = rownames(Y)

Y_list[["BLUP"]] = Y_corrected

# get residuals from modeling Batch as fixed effect
fit = dream(Y, ~ Batch, info, computeResiduals=TRUE)
Y_list[["fixed"]] = mean_per_sampleID(residuals(fit), info)

# get residuals from modeling Batch as random effect
fit = dream(Y, ~ (1|Batch), info, computeResiduals=TRUE)
Y_list[["random"]] = mean_per_sampleID(residuals(fit), info)


# fit = lmer(Y[2,] ~ (1|TimePoint) + (1|Batch) + (1|Individual), info)
# df = data.frame(Expression = Y[5,], info)
# ggplot(df, aes(Individual, Expression)) + geom_point() + theme_bw()
```

# PCA using multiple types of corrections
```{r pca, fig.width=12, fig.height=8}
# Plot coloring by individual
par(mfrow=c(2,3), pty='s')
# PCA of data with no batch effect
dcmp = prcomp( t(Y_clean), scale=TRUE)
plot(dcmp$x[,1:2], col=info$Individual, pch=as.integer(info$Disease)+6, main="Oracle")

# PCA of original data
dcmp = prcomp( t(Y), scale=TRUE)
plot(dcmp$x[,1:2], col=info$Individual, pch=as.integer(info$Disease)+6, main="Observed data")

plot.new()

rs = apply(Y_list[["BLUP"]], 1, sd)
dcmp = prcomp( t(Y_list[["BLUP"]][rs>0,]), scale=TRUE)
idx = match(colnames(Y_list[["BLUP"]]), rownames(metadata))
plot(dcmp$x[,1:2], col=metadata$Individual[idx], pch=as.integer(metadata$Disease[idx])+6, main="Replicate: two step")

Y2 = Y_list[["fixed"]]
rs = apply(Y2, 1, sd)
dcmp = prcomp( t(Y2[rs>0,]), scale=TRUE)
idx = match(colnames(Y2), rownames(metadata))
plot(dcmp$x[,1:2], col=metadata$Individual[idx], pch=as.integer(metadata$Disease[idx])+6, main="Simple v1")

dcmp = prcomp( t(Y_list[["random"]]), scale=TRUE)
idx = match(colnames(Y_list[["random"]]), rownames(metadata))
plot(dcmp$x[,1:2], col=metadata$Individual[idx], pch=as.integer(metadata$Disease[idx])+6, main="Simple v2")
```




```{r differential.expression}

fitList = list()
idx = match(info$SampleID, rownames(metadata))
fitList[['Oracle']] = dream(Y_clean, ~ (1|Disease) + TimePoint + (1|Individual), metadata[idx,])

fitList[['Replicates: shrink batch mean towards equal']] = dream(Y, ~ (1|Disease) + (1|Batch) + TimePoint + (1|Individual), info)
# fitList[['Replicates: set batch mean equal']] = dream(Y, ~ Disease + (1|Batch) + TimePoint + (1|Individual), info)


idx = match(colnames(Y_corrected), rownames(metadata))
rs = apply(Y_corrected, 1, sd)
fitList[['Replicates: Two step']] = dream(Y_corrected[rs > 0,], ~ (1|Disease) + TimePoint, metadata[idx,])
# fitList[['Replicates: Two step']] = eBayes(fitList[['Replicates: Two step']])

idx = ! duplicated(info$SampleID)
fitList[['Naive']] = dream(Y[,idx], ~ TimePoint, info[idx,])
fitList[['Naive']] = eBayes(fitList[['Naive']])


fitList[['Simple']] = dream(Y[,unique_ID_idx], ~ (1|Batch) + TimePoint + (1|Disease), info[unique_ID_idx,])
# fitList[['Simple']] = eBayes(fitList[['Simple']])

res = lapply(names(fitList), function(method){

	tab = topTable(fitList[[method]], coef="TimePointtime2", number=Inf, sort.by="none")

	# add missing genes to tab with P.Value = 1
	excluded = which(!rownames(Y) %in% rownames(tab))

	if( length(excluded) > 0){
		tab_excluded = tab[1:length(excluded),]
		rownames(tab_excluded) = rownames(Y)[excluded]
		tab_excluded[] = NA
		tab_excluded$P.Value = 1
		tab = rbind(tab, tab_excluded)
	}

	idx = rownames(tab) %in% tpGenes

	pr <- pr.curve( -log10(tab$P.Value[idx]), -log10(tab$P.Value[!idx]), curve=TRUE, rand.compute=TRUE )

	data.frame( method 		= method, 
				AUPR 		= pr$auc.integral, 
				AUPR.random = pr$rand$auc.integral,
				stringsAsFactors=FALSE)
	})
res = do.call(rbind, res)

ord = c("Naive", "Simple", "Replicates: Two step", "Replicates: shrink batch mean towards equal", 'Oracle') 
# "Replicates: set batch mean equal",
res$method = factor(res$method, ord)

ggplot(res, aes(method, AUPR, fill=method)) + geom_bar(stat="identity") + theme_bw() + theme(aspect.ratio=1, plot.title = element_text(hjust = 0.5), legend.position="none") + xlab("Method") + scale_y_continuous(limits=c(0, 1.02), expand=c(0, 0)) + geom_hline(yintercept = res$AUPR.random[1], linetype="dashed") + coord_flip() 

```








