# Gabriel Hoffman
# May 20, 2021
#
# Genotype concordance like NGScheckmate, but using genotype dosage
# instead of read depth.  This code works with VCF generated by array and 
# sequencing, while NGScheckmate fails with genotype data
# This code assumes only 1 sample per VCF

# R --args --snpBed /sc/arion/projects/H_PBG/REFERENCES/GRCh38/NGSCheckmate/hglft_genome_6702d_cdce40.bed \
# --vcfList /sc/arion/projects/CommonMind/shan/MOLECULAR_PROFILING/ngscheckmate/final_vcfs.lst \
# --outfile correlation.tsv



library(getopt)

spec = matrix(c(
      'snpBed', 	'd', 1, "character",
      'vcfList',	'v', 1, "character",
      'outfile',	'o', 1, "character"
    ), byrow=TRUE, ncol=4)
opt = getopt(spec)

if( ! file.exists(opt$snpBed) ){
	stop("File does not exist:", opt$snpBed)
}

if( ! file.exists(opt$vcfList) ){
	stop("File does not exist:", opt$snpBed)
}


suppressPackageStartupMessages({
library(VariantAnnotation)
library(GenomicRanges)
library(data.table)
library(parallel)
library(comperes)
library(Repitools)
})

# files = dir("temp", pattern="*.vcf.gz$", full.names=TRUE)

files = read.table(opt$vcfList)$V1

# snpFile = "/sc/arion/projects/H_PBG/REFERENCES/GRCh38/NGSCheckmate/hglft_genome_6702d_cdce40.bed"
gr = with(fread(opt$snpBed), GRanges(V1, IRanges(V2, V3, name=V4, REF=V5, ALT=V6)))
gr = keepSeqlevels(gr, paste0("chr", 1:22), pruning.mode="coarse")
gr$ID = with(gr, paste0(seqnames, ':', end, ':', REF, ':', ALT))

# Compute dosage from genotype probabilities
getDosage = function(genProb){
	sapply(genProb, function(x){
		idx = is.na(x)
		if( length(idx) > 0){
			x[idx] = 0
		}
		# convert probability to dosage
		x[2] + 2*x[3]
		}  )
}

# for each VCF
vcfList = mclapply( files, function(file){

	# read VCF
	res = readVcf( file, genome = "GRCh38" )

	# read variant locations
	gr2 = rowRanges(res)

	# get Dosage for each variant
	if( 'DS' %in% names(geno(res)) ){
		gr2$Dosage = geno(res)$DS
	}else if( "PL" %in% names(geno(res)) ){
		genProb = PLtoGP(geno(res)$PL)
		gr2$Dosage = getDosage( genProb)

		names(gr2) = gsub('_', ':', names(gr2))
		names(gr2) = gsub('/', ':', names(gr2))
	}else if( "GL" %in% names(geno(res)) ){
		genProb = GLtoGP(geno(res)$PL)
		gr2$Dosage = getDosage( genProb)

		names(gr2) = gsub('_', ':', names(gr2))
		names(gr2) = gsub('/', ':', names(gr2))
	}

	# keep only requested variants
	gr2 = gr2[names(gr2) %in% gr$ID]
	attr(gr2, "sample") = samples(header(res))
	gr2
	}, mc.cores=24)

# extract variant info and dosage from each sample
df = lapply( vcfList, function(gr.local){
	res = data.frame(ID=names(gr.local), Dosage = gr.local$Dosage)
	colnames(res)[2] = attr(gr.local, "sample") 
	res
	})

# Merge, keeping all variates
df2 = Reduce(function(x,y) merge(x = x, y = y, by = "ID", all=TRUE), 
       df)

# Compute pairwise correlation
C = cor(df2[,-1], use="pairwise.complete")

# Convert to tall matrix
df_pw = mat_to_long(C, "Sample_1", "Sample_2", "correlation")
df_pw = df_pw[!(df_pw[,1] == df_pw[,2]),]
df_pw$Status = df_pw$correlation > 0.9

# write to file
write.table(df_pw, file=opt$outfile, row.names=FALSE, quote=FALSE, sep="\t")

















