---
title: "COVID and variancePartition"
subtitle: ''
author: "Developed by [Gabriel Hoffman](http://gabrielhoffman.github.io/)"
date: "Run on `r Sys.time()`"
documentclass: article
output: 
  html_document:
  toc: true
  smart: false
---


<!---

# How to run
############

# cd to directory with covid_vp.Rmd 
cd /hpc/users/hoffmg01/work/misc_vp

# set up environment
ml git pandoc 
git pull
R

# in R, pull the latest version (I edit on my laptop and push to GitHub)
# then run the script to generate and HTML page
system("git pull"); rmarkdown::render("covid_vp.Rmd")



--->

```{r setup, include=FALSE}
suppressPackageStartupMessages({
library(ggplot2)
library(data.table)
library(cowplot)
library(kableExtra)
library(knitr)
library(dplyr)
library(compositions)
library(BiocParallel)

library(limma)
library(edgeR)
library(variancePartition)

library(MAST)
library(edgeR)
library(org.Hs.eg.db)
library(limma)
library(lme4)

library(zenith)
})

knitr::opts_chunk$set(
  echo = FALSE,
  warning=FALSE,
  message=FALSE,
  error = FALSE,
  tidy = FALSE,
   dev = c("png", "pdf"),
  cache.lazy = FALSE,
  cache = TRUE)
```


```{r define}
plot_within_btw = function( simMat, design, col=c("lightblue", "lightgreen"),main='',ylim=NULL,...){
  variation = list(type=c(), value=c(), donor1=c(), donor2=c())
  for(i in 1:ncol(design)){
    idx1 = which(design[,i] == 1)   
    idx0 = which(design[,i] == 0)
    C_sub = simMat[idx1, idx1]
    variation$type = append( variation$type, rep("Same Donor", sum(lower.tri(C_sub))) )  
    variation$value = append( variation$value, C_sub[lower.tri(C_sub)])   
    variation$donor1 = append(variation$donor1, rep(colnames(design)[i], length(C_sub[lower.tri(C_sub)]) ))      
    variation$donor2 = append(variation$donor2, rep(colnames(design)[i], length(C_sub[lower.tri(C_sub)]) ))
    variation$type = append( variation$type, rep("Different Donor", length(simMat[idx1,idx0])) )  
    variation$value = append( variation$value, simMat[idx1,idx0])    
    variation$donor1 = append(variation$donor1, rep("", length(simMat[idx1,idx0])) )      
    variation$donor2 = append(variation$donor2, rep("", length(simMat[idx1,idx0])))
  }
  variation = as.data.frame(variation)
  fig = ggplot( variation, aes(type, value)) + geom_violin(scale = "width", aes(fill=type)) + 
    ylab("Variance explained (%)") + xlab("") + geom_boxplot(width = 0.07, 
        fill = "grey", outlier.color = "black") + theme_bw(15) + theme(plot.title = element_text(hjust = 0.5))+ 
        scale_fill_manual(values = col) + theme(legend.position = "none") + 
        ylab("Correlation between samples") + scale_fill_manual(values=c("grey", "steelblue1"))
    if( main !='' ){
      fig = fig + ggtitle( main )
    }
    if( !is.null(ylim) ){
      fig = fig + ylim( ylim )
    }
    fig = fig + theme(aspect.ratio=1)
    return(list(fig=fig, variation=variation))
}
```

# Consider only cell fraction data 
## read data
```{r ct.read}

info = read.table("/sc/arion/projects/CommonMind/roussp01a/COVID_SC/FrozenBrain/results/freq_celltype.csv", sep=',', header=TRUE)
info$covid = factor(info$covid)
info$subject = factor(info$subject)

info$Biosample = factor(with(info, paste(subject, pid, tissue)))

table(info$pid)
table(info$subject)
table(info$clinical)
table(info$tissue)
table(info$covid)
```

# Variance partitioning
Compute variance partitioning for cell fractions (log ratios) for the whole dataset and then each tissue.  The `clr` transformation computes the log ration of each cell type fraction compared to a single, centered value.  The macrophage signal is only apparent with this transformation.

### Hierarchical clustering
```{r CT.hclust, fig.width=16}
# use observed cell type frations
# CT.quant = t(info[,1:15])

# use centered log ratios
pseudoFrac = 1e-2
CT.quant = t(clr(info[,1:15] + pseudoFrac))

D = 1 - cor(CT.quant)
colnames(D) = info$Biosample
hcl = hclust(as.dist(D))
plot(hcl)
colnames(CT.quant) = 1:ncol(CT.quant)
```

### PCA
```{r pca, fig.width=12, fig.height=6}
dcmp = prcomp( t(CT.quant) )
info = cbind(info, dcmp$x[,1:4])
rownames(info) = c()

ggplot( info, aes(PC1, PC2, color=subject, shape=tissue)) + geom_point() + theme_bw() + theme(aspect.ratio=1)
```

### Compare similarity between samples from same or different donors for each brain region
It seems interesting that there is so much variation *across* donors in the choroidpleus compared to PFC.  Is that biological or technical?
```{r, within.between, fig.width=12, fig.height=6}
figList = lapply( sort(unique(info$tissue)), function(Tissue){
  idx = which(info$tissue == Tissue)
  C = cor(CT.quant, method="spearman")
  dsgn = model.matrix(~0+subject, info[idx,])

  res = plot_within_btw( C[idx,idx], dsgn)

  wt = with( unique(res$variation), wilcox.test( value[type=="Same Donor"], value[type=="Different Donor"], alternative="greater"))

  res$fig + ggtitle( paste0(Tissue, ": (p=", format(wt$p.value, digits=3), ')') ) + ylim(-0.5, 1)
})
plot_grid( plotlist=figList, nrow=1)
```





```{r cancor, fig.width=6, fig.height=6, eval=FALSE}
C = canCorPairs(~ pid + subject + covid + tissue + Biosample, info)

plotCorrMatrix(C)
```

There are 4 samples per subject per brain region.  These 4 samples represent 2 disections and 2 10X runs.  I refer to the same `Biosample`  to mean to same physical piece of tissue run on two 10X batches.    So even though there are 108 samples, there are only 9 subjects. Based on the study design, and the variance partition below it is essential to account for `(1|subject)`.  Since the 10X run accounts for very little variation, it is not critical to account for `(1|Biosample)` but the results dont change much if you do.  

This is exactly the problem that `dream` (by calling `lmer`) is meant to address.  Ignoring the repeated measures can lead to false positives.  See [Jostins, et al, PNAS 2012](https://www.pnas.org/content/109/18/E1048) for an example. 


```{r facet_wrap, fig.width=12, fig.height=30}
df  = cbind(info[,16:22], t(CT.quant))
df_melt = reshape2::melt(df, id.vars=colnames(info[,16:22]))

ggplot(df_melt, aes(subject, value, color=covid)) + geom_point() + theme_bw() + theme(aspect.ratio=1, axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) + facet_wrap(~ variable + tissue, ncol=3) + ylim('Centered log ratio')
```

### Variance partition for the entire dataset

```{r CT.vp}
# form = ~ (1|pid) + (1|subject) + (1|covid) + (1|tissue) 
form = ~ (1|pid) + (1|subject) + (1|covid) + (1|tissue) + (1|Biosample)

vp = fitExtractVarPartModel( CT.quant, form, info)   

plotPercentBars(vp)
``` 

### Variance partition for each tissue
```{r per.tissue, fig.width=12}
form = ~ (1|pid) + (1|subject) + (1|covid) + (1|Biosample)
vpList = lapply( unique(info$tissue), function(Tissue){

  idx = which(info$tissue == Tissue)
 
  fitExtractVarPartModel( CT.quant[,idx], form, info[idx,]) 
})
names(vpList) = unique(info$tissue)

figList = lapply( names(vpList), function(id){
  plotPercentBars( vpList[[id]] ) + ggtitle(id) + theme(legend.position="bottom")
  })   

plot_grid(plotlist = figList, ncol=3)
```

### Hypothesis testing for each tissue
Omitting `(1|subject)` produces a lot of positive findings because the model sees 36 samples, but doesnt consider the correlation structure.

```{r per.tissue.dream, fig.width=12}
form = ~ (1|pid) + (1|subject) + covid + (1|Biosample)

tissues = sort(unique(info$tissue))
fitList = lapply( tissues, function(Tissue){

  idx = which(info$tissue == Tissue)
 
  cellFrac = data.frame(CT.quant[,idx], check.names=FALSE)

  dream( cellFrac, form, info[idx,]) 
})
names(fitList) = tissues


tabList = lapply( tissues, function(Tissue){

  fit = fitList[[Tissue]]

  df = topTable(fit, coef="covid1", sort.by='none', number=Inf)

  df$Tissue = Tissue
  df$CT = rownames(df)

  df
})
tab = do.call(rbind, tabList)

tab$FDR = p.adjust(tab$P.Value, "fdr")

ggplot(tab, aes(CT, z.std, fill=(FDR < 0.10))) + geom_bar(stat="identity") + facet_wrap(~Tissue) + theme_bw() + theme(aspect.ratio=1) + coord_flip() + scale_fill_manual(values=c("black", "red"))
```

```{r joint.model, results="asis"}
# test combined covid:tissue
form = ~ covid*tissue + (1|subject) + (1|Biosample)

Y = data.frame(CT.quant, check.names=FALSE)
fit = dream( Y, form, info, BPPARAM=SerialParam()) 
  
topTable(fit, coef="covid1") %>% kable() %>% kable_styling(full_width=FALSE)
```


# Gene expression
```{r define.functions}

write_csv = function(vec, fn){
  write.table(vec, file = fn, append = FALSE, quote = FALSE, sep = ",",
                eol = "\n", na = "NA", dec = ".", row.names = TRUE,
                col.names = TRUE, qmethod = c("escape", "double"),
                fileEncoding = "")
}

prepare_sca = function(expr){
    ncells = nrow(expr)
    ngenes = ncol(expr)   
    
    fData <- data.frame(primerid=colnames(expr))
    rownames(fData) = colnames(expr)
    cData <- data.frame(wellKey=rownames(expr))
    rownames(cData) = rownames(expr)

    mat <- as.matrix(t(expr))
    normed_mat = log2(1 + cpm.default(mat))

    sca <- FromMatrix(normed_mat, cData, fData)
    sca
}

get_hurdel_result = function(summaryDt){
    to.return <- as.data.frame(subset(summaryDt, component == "H")[, c(1, 4)])
    colnames(to.return) = c('gene', 'pval')
    to.return$adjp = p.adjust(to.return$pval)
    rownames(to.return) = to.return$gene
    to.return = to.return[, c(2, 3)]
    
    to.return
}
```


```{r read.gene.expression, cache=FALSE}
library(dplyr)
library(Matrix)
geneExpr = fread('/sc/arion/projects/CommonMind/roussp01a/COVID_SC/FrozenBrain/results/gene_expression_v10k.csv') %>% as.data.frame
rownames(geneExpr) = geneExpr$V1
geneExpr = geneExpr[,-1]
geneExpr = as.matrix(geneExpr)

sca = prepare_sca( geneExpr)

metadata = fread('/sc/arion/projects/CommonMind/roussp01a/COVID_SC/FrozenBrain/results/metadata.csv')

colData(sca)$subject_id <- factor(metadata$subject_id)
colData(sca)$tissue <- factor(metadata$tissue)
colData(sca)$pid <- factor(metadata$pid)  
colData(sca)$celltype <- factor(metadata$celltype)
colData(sca)$is_covid <- metadata$is_covid  
colData(sca)$Biosample = factor(with(metadata, paste(subject_id, pid, tissue)))
```


### SCTransform from SEURAT
I used sctransform here because it is been widely adapted.   I think I could have used `MAST::zlm`, but it was more trouble than it was worth.
```{r sctransform}
library(sctransform)
vst_out <- sctransform::vst( t(geneExpr), latent_var = c("log_umi"), return_gene_attr = TRUE, 
    return_cell_attr = TRUE, verbosity = 1)
```

Analysis of entire dataset.  In order to make this computationally tractible, I summarize the expression all cells from a given Biosample in a single cell type in terms of the mean.  The reduces the regression from 68K cells to 761 Biosample/celltype combinations.
```{r vp.sc}
metadata$Biosample = factor(with(metadata, paste(subject_id, pid, tissue)))
metadata$ID = factor(with(metadata, paste(subject_id, pid, tissue, celltype)))

geneExpr_mean = lapply( levels(metadata$ID), function(id){
  idx = which(metadata$ID == id)

  y = t(t(rowMeans(vst_out$y[,idx,drop=FALSE])))
  colnames(y) = id
  y
  })
geneExpr_mean = do.call(cbind, geneExpr_mean)

metadata_mean = with(metadata, unique(data.frame(subject_id, pid, tissue, celltype, Biosample, ID, clinical, is_covid)))

idx = match( colnames(geneExpr_mean), metadata_mean$ID)
metadata_mean = metadata_mean[idx,]

form = ~ (1|pid) + (1|subject_id) + (1|clinical) + (1|tissue) + (1|Biosample) + (1|celltype)

vp = fitExtractVarPartModel( geneExpr_mean, form, metadata_mean, BPPARAM=SnowParam(36, progressbar=TRUE))   
```

```{r plot.vp.summary}
plotVarPart( sortCols(vp) )
```

# Analysis within each cell type
You probably want to add an additional analysis within brain region 
```{r vp.CT}
form = ~ (1|pid) + (1|subject_id) + (1|clinical) + (1|tissue)

CTs = unique(metadata_mean$celltype)
vpList = lapply( CTs, function(CT){

  idx = which(metadata_mean$celltype == CT)

  fitExtractVarPartModel( geneExpr_mean[,idx], form, metadata_mean[idx,], BPPARAM=SnowParam(15, progressbar=TRUE))   
  })
names(vpList) = CTs
```

```{r plot.vp.CT.summary, fig.width=8, fig.height=20}
figList = lapply( names(vpList), function(ID){
  vp = vpList[[ID]]
  plotVarPart( sortCols(vp) ) + ggtitle( ID)
  })
plot_grid(plotlist=figList, ncol=2)
```

# Count aggregation

```{r aggregate}

aggregate_counts = function( geneExpr, metadata){

  # aggrecate gene counts
  geneCounts = lapply( levels(metadata$ID), function(id){

      idx = which(metadata$ID == id)

      colSums(geneExpr[idx,,drop=FALSE])
    })
  geneCounts = do.call(cbind, geneCounts)
  colnames(geneCounts) = levels(metadata$ID)

  # simplify metadata 
  md2 = lapply( levels(metadata$ID), function(id){

      unique(metadata[metadata$ID == id,c("subject_id", "ID", "Biosample", "clinical", "is_covid", "pid", "tissue", "celltype")])
  })
  md2 = do.call(rbind, md2)

  list( geneCounts  = geneCounts,
        metadata    = md2) 
}

res = aggregate_counts( geneExpr, metadata)


dge = DGEList( res$geneCounts, samples=res$metadata )

dge = calcNormFactors(dge)

dsgn = model.matrix(~ tissue + celltype + subject_id, dge$samples )
vobj = voom(dge, dsgn, plot=TRUE)
```

## variancePartition
```{r vp.aggregate}
form = ~ (1|tissue) + (1|celltype) + (1|tissue:celltype) + (1|is_covid) + (1|subject_id) + (1|Biosample)

vp = fitExtractVarPartModel(vobj, form, dge$samples, BPPARAM=SnowParam(3, progressbar=TRUE))
```

```{r plot.vp.summary.aggregate}
plotVarPart( sortCols(vp) )
```

## dream
```{r dream.aggregate}
form = ~ tissue*celltype*is_covid + (1|subject_id) + (1|Biosample)

fit = dream(vobj, form, dge$samples, BPPARAM=SnowParam(3, progressbar=TRUE))
```




<!---

for(CT in unique(vSMC$celltype)){

  idx = which(res$metadata$celltype == CT)

  dge = DGEList( res$geneCounts[,idx], samples=res$metadata[idx,] )

  dge = calcNormFactors(dge)

  dsgn = model.matrix(~ subject_id, dge$samples )
  vobj = voom(dge, dsgn, plot=TRUE)



--->







































