---
title: "Impute z-scores from ungenotyped variants"
subtitle: 'Using observed z-scores and LD matrix'
author: "Developed by [Gabriel Hoffman](http://gabrielhoffman.github.io/)"
date: "Run on `r Sys.time()`"
output: 
  html_document:
    toc: false
    smart: true
params:
  DATASET: NULL
---

<!----


rmarkdown::render("impute_z.Rmd")


--->

# Main function
```{r imputed_z}
#' Impute unobserved z-scores given observed data
#'
#' Impute z-scores at unobserved variants using z-scores and LD matrix from observed variants.
#' 
#' @param z array of z-scores
#' @param Sigma LD matrix 
#' @param i index to impute
#' @param lambda tuning parameter for shrinking covariance matrix
#' 
#' @description
#' Implement method from Pasaniuc, et al. (Bioinformatics, 2014) using a multivariate approximation of the observed z-scores
#' 
#' @return
#' data.frame storing variant ID, imputed z-statistic (z.stat) and standard error of the imputed z-score (se).
#' 
impute_z = function(z, Sigma, i, lambda = 0.1){

	if( is(Sigma, "sparseMatrix") ){
		Sigma.shrink = (1-lambda) * Sigma + Diagonal(nrow(Sigma), lambda)
	}else{
		Sigma.shrink = (1-lambda) * Sigma + diag(lambda, nrow(Sigma))
	}

	# impute the ith z-score using 
	# impute z-score using Gaussian conditional distribution
	# z_i = Sigma.shrink[i,-i] %*% solve(Sigma.shrink[-i,-i], z[-i])
	# W = Sigma.shrink[i,-i] %*% solve(Sigma.shrink[-i,-i])

	# weights
	W = solve(Sigma.shrink[-i,-i], Sigma.shrink[-i,i,drop=FALSE])

	# imputed z-scores
	z_i = crossprod(W, z[-i])

	# compute standard error for each imputed z-score
	se_i = sqrt(diag(crossprod(W,Sigma.shrink[-i, -i]) %*% W))

	data.frame(ID = names(z)[i], 
						z.stat = as.numeric(z_i), 
						se = as.numeric(se_i))
}
```

# Simple simulation
```{r analysis, fig.width=4.5}
library(Rfast)

# Generate auto-correlation matrix
autocorr.mat = function (p, rho){
    mat <- diag(p)
    rho^abs(row(mat) - col(mat))
}

# Simulate LD matrix
p = 500
rho = .95
Sigma = autocorr.mat(p, rho )

# Simulate z-scores
z = rmvnorm(1, rep(0, p), Sigma, seed=1)
z = as.numeric(z)
names(z) = paste0("SNP_", 1:p)

# Plot z-scores
plot(z, xlab="Variants", ylab="z-scores")
```

# Impute the ith variant
```{r test}
i = 50
impute_z(z, Sigma, i)
```

# Impute each variant 
For the ith variant, use observations from all other variants.
```{r all}
df_z = lapply(1:p, function(i){
	impute_z(z, Sigma, i)
})
df_z = do.call(rbind, df_z)

head(df_z)
```

# Plot each variant
```{r plot, fig.width=4.5}
# Plot comparing observed and imputed values
plot(z, df_z$z.stat, xlab="Observed z", ylab="Imputed z", main="Compare observed and imputed")
abline(0, 1, col="red")
r = cor(z, df_z$z.stat)
text(x=0, y=1.8, paste0("R = ", format(r, digits=3)))
```

# Read from TOPLD
```{r topLD}
library(tidyverse)
library(Matrix)

# LD filtered at R2 > .2, sp R > 0.44
df = read_csv("~/Downloads/EUR_chr22_no_filter_0.2_1000000_LD.csv.gz") %>%
	filter(SNP1 < SNP2) %>%
	rename( Sign = `+/-corr`) %>%
	mutate( Sign = ifelse(Sign=='+', 1, -1))

# Get unique positions, sorted increasing order
df_position = rbind(df %>% 
					select(SNP1, Uniq_ID_1) %>% 
					rename(Position = SNP1, ID = Uniq_ID_1), 
					df %>% 
					select(SNP2, Uniq_ID_2) %>% 
					rename(Position = SNP2, ID = Uniq_ID_2)) %>%
					distinct %>%
					arrange(Position)

N = nrow(df_position)

# Match SNPs to order positions
df$idx.i = match(df$Uniq_ID_1, df_position$ID)
df$idx.j = match(df$Uniq_ID_2, df_position$ID)

# Create sparse matrix storing LD
M = sparseMatrix(i = df$idx.i,
				 j = df$idx.j,
				 x = with(df, sqrt(R2)*Sign),
				 symmetric = TRUE,				 
				 dims = c(N,N), 
				 dimnames=list(df_position$ID, df_position$ID) )
diag(M) = 1

sparsity = function(x){
	1 - 2*length(x@x) / prod(x@Dim)
}

sparsity(M)

# save M and df_position for each chromsome
```

## Simulate using real LD
The LD filter is very strictt R2 > 0.4.  This means that very few non-zero values are reported and the LD matrix may not be positive definite.  This makes the operation very fast, but it means that this LD resource will give less than ideal accuracy of imputed z-scores.  You will likely need a more complete LD matrix for this to be useful
```{r sim2}
# library(sparseMVN)
# number of SNPs
p = 1000 
i = seq(1, p)

lambda = 0
Sigma = (1-lambda)*M[i,i] + lambda*Diagonal(p, 1)

Sigma = as(autocorr.mat(p, rho ), "sparseMatrix")

ch = Cholesky(Sigma)
A = Matrix::expand(ch)

# Simulate z-scores
z = crossprod(A$P, A$L %*% rnorm(p))
z = as.numeric(z)
names(z) = paste0("SNP_", 1:p)

# Plot z-scores
plot(z, xlab="Variants", ylab="z-scores")
```

# Impute the ith variant
```{r test.sparse}
i = 50
impute_z(z, Sigma, i)
```

# Impute each variant 
For the ith variant, use observations from all other variants.
```{r all.sparse}
df_z = lapply(1:p, function(i){
	impute_z(z, Sigma, i)
})
df_z = do.call(rbind, df_z)

head(df_z)
```

# Plot each variant
```{r plot.sparse, fig.width=4.5}
# Plot comparing observed and imputed values
plot(z, df_z$z.stat, xlab="Observed z", ylab="Imputed z", main="Compare observed and imputed")
abline(0, 1, col="red")
r = cor(z, df_z$z.stat)
text(x=0, y=1.8, paste0("R = ", format(r, digits=3)))
```



