# Gabriel Hoffman
# April 26, 2023


library(data.table)
library(Matrix)
library(GenomicRanges)

#' Read LD LD files 
#'
#' Read LD LD files generated by plink into a list of sparseMatrix for each chromosome 
#'
#' @param df_files data.frame of file info: chrom, BIM, LD
#' @param r2cutoff values with r^2 less than this are set to zero
readLDMatrix = function( df_files, r2cutoff=0.05){

	# for each chromosome
	r2MatList = lapply( seq_len(nrow(df_files)), function(i){

		chrom = df_files$chrom[i]
		cat("\r", chrom, '   ')

		# read marker position
		df_position = fread( df_files$BIM[i] )
		colnames(df_position) = c("chrom", "name", "gmap", "position", "allele1", "allele2")

		# read LD
		# this does R-squared
		cmd = paste0("zcat ", df_files$LD[i], " | awk '{if($7*$7 > ", r2cutoff, ") print $3, $6, $7}'")
		dfld = fread( cmd=cmd, sep=' ')
		colnames(dfld) = c('SNP_A', 'SNP_B', "R")

		# keep only variants in dfld
		df_position = df_position[name %in% unique(c(dfld$SNP_A, dfld$SNP_B)),]

		# create GRanges object for variants in LD matrix
		gr_position = with(df_position, GRanges(paste0('chr',chrom), IRanges(position, position, name=name), allele1=allele1, allele2=allele2))

		# Add self correlation of 1 
		df_self = data.frame(	SNP_A = df_position$name, 
								SNP_B = df_position$name,
								R = 1,
								stringsAsFactors=FALSE)
		dfld = rbind(dfld, data.table(df_self))

		# get index for each marker
		dfld$idx_A = match(dfld$SNP_A, df_position$name)
		dfld$idx_B = match(dfld$SNP_B, df_position$name)

		N = nrow(df_position)

		# Create sparse matrix
		C = sparseMatrix( i = dfld$idx_A,
						j = dfld$idx_B,
						x = dfld$R,
						symmetric = TRUE, 
						dims = c(N,N), 
						dimnames=list(df_position$name, df_position$name) )
		rm(dfld)

		# return LD matrix and variant annotation
		list(C.ld = C, 
			gr = gr_position)
	})

	names(r2MatList) = as.character(df_files$chrom)

	r2MatList
}

fillRate = function( M ){
	if( isSymmetric(M) ){
		rate = 2 * length(M@x) / nrow(M)^2
	}else{
		rate = length(M@x) / (nrow(M) * ncol(M))
	}
	rate
}



# This example uses a data frame with 1 row
# for real analysis use rows for 1:22
df = data.frame(chrom = 22)
df$BIM = "/sc/arion/projects/roussp01a/sanan/230322_GWASimp/imputeZPipeline_V2/inter/230424/plinkStep2/1kg_chr22.bim"

df$LD = "/sc/arion/projects/roussp01a/sanan/230322_GWASimp/imputeZPipeline_V2/inter/230424/plinkStep3/1kg_ld_chr22.ld.gz"

LDm = readLDMatrix( df, r2cutoff=0.01 )

LDm[["22"]]$C.ld[1:3, 1:3]
LDm[["22"]]$gr[1:3]

length(LDm[["22"]]$gr)


fillRate( LDm[["22"]]$C.ld )

saveRDS(LDm, file="LDm.RDS")

LDm = readRDS("LDm.RDS")





