---
title: "OneK1K single cell RNA-seq"
subtitle: 'https://onek1k.org'
author: "Developed by [Gabriel Hoffman](http://gabrielhoffman.github.io/)"
date: "Run on `r Sys.time()`"
output: 
  html_document:
    toc: true
    smart: true
    self_contained: true
---

<!---


cd /hpc/users/hoffmg01/work/misc_vp
R --vanilla

system("git pull"); rmarkdown::render("yazar_2022.Rmd");


# https://hoffmg01.hpc.mssm.edu/crumblr_analysis/Yazar_Science_2022/yazar_2022.html



# cd 

--->


```{r setup, include=FALSE}
knitr::opts_chunk$set(
  echo = TRUE,
  warning=FALSE,
  message=FALSE,
  error = FALSE,
  tidy = FALSE,
  dev = c("png", "pdf"),
  package.startup.message = FALSE,
  cache = TRUE,
  cache.lazy = FALSE)
```


```{r load, cache=FALSE}
library(SingleCellExperiment)
library(zellkonverter)
library(dreamlet)
library(tidyverse) 
library(ggplot2)
```

```{r read.data}
# read single cell RNA-seq
file = "/sc/arion/projects/CommonMind/hoffman/scRNAseq_data/Yazar_Science_2022/08984b3c-3189-4732-be22-62f1fe8f15a4.h5ad"
sce = readH5AD(file, use_hdf5=TRUE)
counts(sce) = assay(sce, "X")

# Compute pseudobulk by donor
pb <- aggregateToPseudoBulk(sce,
    assay = "counts",     
    cluster_id = "cell_type",  
    sample_id = "donor_id",
    verbose = FALSE)
rm(sce)
```


```{r voom}
param = SnowParam(12)

# form = ~ age + (1|sex) + (1|pool_number) 
form = ~ age + sex

res.proc <- processAssays(pb, form, BPPARAM=param, min.cells=1, min.count=1, min.prop=0.1)

res.proc

details(res.proc)
```

```{r voom.plot, fig.width=7, fig.height=20}
plotVoom(res.proc)
```


```{r strict}
res.proc.strict <- processAssays(pb, form)
```

```{r fitVarPart}
vp.lst <- fitVarPart(res.proc, form, BPPARAM=param)
```

```{r plotVarPart, fig.width=7, fig.height=20}
plotVarPart(vp.lst, label.angle = 60)
```

```{r dreamlet}
res.dl <- dreamlet(res.proc, form, BPPARAM=param)
```

```{r plotVolcano, fig.width=7, fig.height=20}
plotVolcano(res.dl, coef = "age", nGenes=1, scales = "free_y") 
```

The fraction of 'naive thymus-derived CD8-positive, alpha-beta T cell' drops from 20% to <0.1%.  This is a dramatic change.
```{r cell.freq}
counts = cellCounts(pb)
fracs = counts / rowSums(counts)

cellType = "naive thymus-derived CD8-positive, alpha-beta T cell"

par(pty="s")
plot( 100*fracs[,cellType] ~ pb$age, ylab="Cell percentage (CD8+, alpha-beta T cell) [log10 scale]", xlab="Age", log='y')

nReads = colSums(assay(pb, cellType))
plot( 100*fracs[,cellType] ~ nReads, ylab="Cell percentage (CD8+, alpha-beta T cell) [log10 scale]", xlab="nReads", log='xy')

plot( nReads ~ pb$age, xlab="Age", log='y')
```

df_merge




```{r topTable}
tab = topTable(res.dl, coef="age", number=Inf) %>%
  as_tibble %>%
  group_by(assay) %>%
  summarize(nDown = sum(adj.P.Val < 0.05 & logFC < 0), nUp = sum(adj.P.Val < 0.05 & logFC > 0), nGenes = length(adj.P.Val)) %>%
  arrange(nGenes) %>%
  data.frame

ggplot(tab) +
  geom_bar(aes(assay, nUp), stat="identity", fill="red3") + 
  geom_bar(aes(assay, -nDown), stat="identity", fill="blue3") + 
  theme_classic() +
  theme(aspect.ratio=1) +
  ylab("# DEGs") +
  coord_flip() 
```


```{r focus}
cellType = "naive thymus-derived CD8-positive, alpha-beta T cell"

topTable(res.dl, coef="age", number=Inf) %>%
  as_tibble %>%
  filter(assay == cellType) %>%
  head %>%
  data.frame

df_merge <- extractData(res.proc, cellType)

par(pty="s")
plot(ENSG00000187514 ~ age, df_merge)
```

```{r strict.results}
df_merge_strict <- extractData(res.proc.strict, cellType)

dim(df_merge)
dim(df_merge_strict)

genes_strict = rownames(assay(res.proc.strict, cellType))


tab = topTable(res.dl, coef="age", number=Inf) %>%
  as_tibble %>%
  filter(ID %in% genes_strict) %>%
  group_by(assay) %>%
  summarize(nDown = sum(adj.P.Val < 0.05 & logFC < 0), nUp = sum(adj.P.Val < 0.05 & logFC > 0), nGenes = length(adj.P.Val)) %>%
  arrange(nGenes) %>%
  data.frame

ggplot(tab) +
  geom_bar(aes(assay, nUp), stat="identity", fill="red3") + 
  geom_bar(aes(assay, -nDown), stat="identity", fill="blue3") + 
  theme_classic() +
  theme(aspect.ratio=1) +
  ylab("# DEGs") +
  coord_flip() 

 topTable(res.dl, coef="age", number=Inf) %>%
  as_tibble %>%
  filter(! ID %in% genes_strict) %>%
  filter(assay == cellType) %>%
  head %>%
  data.frame

# Gene increase with age in liberal, but not included in strict
geneID = "ENSG00000128309"

par(pty="s")
plot(df_merge$age, df_merge[[geneID]])
# plot(ENSG00000128309 ~ age, df_merge_strict)

nGeneReads = assay(pb, cellType)[geneID,]
plot(pb$age, nGeneReads)

nReads = colSums(assay(pb, cellType))
plot(pb$age, log2((nGeneReads+.25) / (nReads/1e6)))
```






```{r voom2}

df_cells = cellCounts(pb) %>%
  as.data.frame %>%
  rownames_to_column("donor_id") %>%
  as_tibble %>%
  pivot_longer(!donor_id) %>%
  rename(nCells = value, cell_type=name)

metadata(pb)$aggr_means = metadata(pb)$aggr_means %>%
    left_join(df_cells)

param = SnowParam(12)

# form = ~ age + (1|sex) + (1|pool_number) 
form = ~ age + sex + log(nCells)

res.proc <- processAssays(pb, form, BPPARAM=param)

res.proc

details(res.proc)
```

```{r voom.plot2, fig.width=7, fig.height=20}
plotVoom(res.proc)
```


```{r strict2}
res.proc.strict <- processAssays(pb, form)
```

```{r fitVarPart2}
vp.lst <- fitVarPart(res.proc, form, BPPARAM=param)
```

```{r plotVarPart2, fig.width=7, fig.height=20}
plotVarPart(vp.lst, label.angle = 60)
```

```{r dreamlet2}
res.dl <- dreamlet(res.proc, form, BPPARAM=param)
```

```{r plotVolcano2, fig.width=7, fig.height=20}
plotVolcano(res.dl, coef = "age", nGenes=1, scales = "free_y") 
```



